src/kernels/MIOpenArgmax.cpp

```diff
@@ -0,0 +1,61 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+
+extern "C" __global__ void ArgmaxFwdContiguous(const FLOAT* __restrict__ x,
```

Comment:
Perhaps we can convert this to a device function templated on the data type so that a new kernel does not need to be compiled for each data type.

Reply:
When I adding the HIP kernel for the first time, the kernel was created by referring to https://github.com/ROCm/MIOpen/blob/dfaff6aa6077b19e34e7b15c0a74fb055bb6256d/src/kernels/MIOpenTensorKernelsHip.cpp#L53. Can you please provide another example using template?

Reply:
Please see: https://github.com/ROCm/MIOpen/blob/e4f1c8e7c2ab2b6c5e41d5371d9fa0f1113f389d/src/kernels/gpu_reference_kernel/naive_conv.cpp

General patterns is that we define an `extern "C" __global__ void kernel()` for each data-type but all such kernels call a templated `__device__ function` that implements a generic design.

Reply:
There seem to be two ways to use template funtion. 
The first is to create all global functions using #define and call the template device function accordingly. (https://github.com/ROCm/MIOpen/blob/e4f1c8e7c2ab2b6c5e41d5371d9fa0f1113f389d/src/kernels/gpu_reference_kernel/naive_conv.cpp)
Second, when building the kernel, it receives #define INPUT_TYPE, builds a global function, and calls the template device function accordingly. (https://github.com/ROCm/MIOpen/blob/e4f1c8e7c2ab2b6c5e41d5371d9fa0f1113f389d/src/kernels/gpu_reference_kernel/fp8_naive_conv.cpp)
The second method has been recently updated and seems suitable to apply to argmax.
Will there be any problems if I proceed with the second method?

Reply:
Either one is an OK choice.

Reply:
Making it device template unfortunately doesn't help much.
In that case we have to do dynamic dispatch and compile all the types in any case (even if we don't use some of them).

Reply:
@seungmanhan We do not use the conventional HIP calling machinery (i.e. normally we do not directly call device code from the host code, -- and this solver too), so please do not use `__global__`.